sql/core/src/test/scala/org/apache/spark/sql/PlanStabilitySuite.scala

```diff
@@ -0,0 +1,335 @@
+      case _ => -1
+    }
+
+    /**
+     * Some expression names have ExprId in them due to using things such as
+     * "sum(sr_return_amt#14)", so we remove all of these using regex
+     */
+    def cleanUpReferences(references: AttributeSet): String = {
+      val rgxId = "#\\d+".r
+      rgxId.replaceAllIn(references.toSeq.map(_.name).sorted.mkString(","), "")
+    }
+
+    /**
+     * Generate a simplified plan as a string
+     * Example output:
+     * TakeOrderedAndProject [c_customer_id]
+     *   WholeStageCodegen
+     *     Project [c_customer_id]
+     */
+    def getSimplifiedPlan(node: SparkPlan, depth: Int): String = {
+      val padding = "  " * depth
+      var thisNode = node.nodeName
+      if (node.references.nonEmpty) {
+        thisNode += s" [${cleanUpReferences(node.references)}]"
+      }
+      if (node.producedAttributes.nonEmpty) {
+        thisNode += s" [${cleanUpReferences(node.producedAttributes)}]"
+      }
+      val id = getId(node)
+      if (id > 0) {
+        thisNode += s" #$id"
+      }
+      val childrenSimplified = node.children.map(getSimplifiedPlan(_, depth + 1))
+      val subqueriesSimplified = node.subqueries.map(getSimplifiedPlan(_, depth + 1))
+      s"$padding$thisNode\n${subqueriesSimplified.mkString("")}${childrenSimplified.mkString("")}"
+    }
+
+    getSimplifiedPlan(plan, 0)
+  }
+
+  private def normalizeIds(query: String): String = {
+    val regex = "#\\d+L?".r
+    val map = new mutable.HashMap[String, String]()
+    regex.findAllMatchIn(query).map(_.toString)
+      .foreach(map.getOrElseUpdate(_, (map.size + 1).toString))
+    regex.replaceAllIn(query, regexMatch => s"#${map(regexMatch.toString)}")
+  }
+
+  /**
+   * Test a TPC-DS query. Depending on the settings this test will either check if the plan matches
+   * a golden file or it will create a new golden file.
+   */
+  protected def testQuery(tpcdsGroup: String, query: String, suffix: String = ""): Unit = {
+    val queryString = resourceToString(s"$tpcdsGroup/$query.sql",
+      classLoader = Thread.currentThread().getContextClassLoader)
+    val qe = sql(queryString).queryExecution
+    val plan = qe.executedPlan
+    val explain = normalizeIds(qe.toString)
+
+    if (regenerateGoldenFiles) {
+      generateApprovedPlanFile(plan, query + suffix, explain)
+    } else {
+      checkWithApproved(plan, query + suffix, explain)
+    }
+  }
+
+  // TODO: this is used to reduce generated files to ease the review. Revert it back after review.
+  val targetQuires = Set("q47", "q49", "q57")
+}
+
+@ExtendedSQLTest
+class TPCDSV1_4_PlanStabilitySuite extends PlanStabilitySuite {
+  override val goldenFilePath: String =
+    new File(baseResourcePath, s"approved-plans-v1_4").getAbsolutePath
+
+  tpcdsQueries.filter(targetQuires.contains).foreach { q =>
+    test(s"check simplified (tpcds-v1.4/$q)") {
+      testQuery("tpcds", q)
+    }
+  }
+}
+
+@ExtendedSQLTest
```

Comment:
Have you checked the running time? https://github.com/apache/spark/pull/29270/files#r469675879

Reply:
Ah..I don't have a chance to do that yet. I could do it later. Actually, there's a more critical problem that relates to the failed tests, that is,

running the following two commands, q49 can succeed with cmd (1) but fail with cmd(2), which means q49 generates different plans with these 2 two commands. I'm still looking at this problem.

```
build/sbt "sql/test-only *TPCDSV1_4_PlanStabilitySuite -- -z tpcds-v1.4"  (1)
build/sbt "sql/test-only *TPCDSV1_4_PlanStabilitySuite -- -z tpcds-v1.4/q49" (2)
```